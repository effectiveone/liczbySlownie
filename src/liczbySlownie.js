import { odmianaZlotych } from "./odmianazlotych.js";
import { groszeSlownie } from "./groszeslownie.js";
import {
  grupy,
  setki,
  dziesiatki,
  nascie,
  jednosci,
} from "./odmianadziesiatek.js";

export function liczbySlownie(liczba) {
  const licznik = liczba != null ? liczba.toFixed(2) : null;
  let [zloteRaw, groszeRaw] = licznik.toString().split(".");
  let zlote = parseInt(zloteRaw);
  const nominal = odmianaZlotych(zlote);
  let grosze = parseInt(groszeRaw || "0");

  if (isNaN(zlote) || isNaN(grosze)) {
    return "Niepoprawna wartość";
  }

  let wynik = "";
  let znak = "";
  if (zlote === 0 && grosze === 0) {
    wynik = "zero";
  }
  if (zlote < 0) {
    znak = "minus";
    zlote = -zlote;
  }

  let g = 0;
  while (zlote > 0) {
    let s = Math.floor((zlote % 1000) / 100);
    let n = 0;
    let d = Math.floor((zlote % 100) / 10);
    let j = Math.floor(zlote % 10);

    if (d == 1 && j > 0) {
      n = j;
      d = 0;
      j = 0;
    }

    let k = 2;
    if (j == 1 && s + d + n == 0) {
      k = 0;
    }
    if (j == 2 || j == 3 || j == 4) {
      k = 1;
    }
    if (s + d + n + j > 0) {
      wynik =
        setki[s] +
        dziesiatki[d] +
        nascie[n] +
        jednosci[j] +
        grupy[g][k] +
        wynik;
    }

    g++;
    zlote = Math.floor(zlote / 1000);
  }

  if (grosze === 0) {
    return wynik.trim() + " " + nominal;
  }

  let wynikGrosze = "";

  if (grosze === 1) {
    wynikGrosze = "jeden grosz";
  } else if (
    grosze % 10 >= 2 &&
    grosze % 10 <= 4 &&
    Math.floor(grosze / 10) !== 1
  ) {
    wynikGrosze = `${groszeSlownie(grosze)} grosze`;
  } else {
    wynikGrosze = `${groszeSlownie(grosze)} groszy`;
  }

  if (wynik === "") {
    return znak + wynikGrosze.trim();
  }

  return `${wynik}${nominal} i ${wynikGrosze}`;
}

// przykłady użycia
console.log(liczbySlownie(123.45));
console.log(liczbySlownie(500));
console.log(liczbySlownie(9999.99));
console.log(liczbySlownie(0.01));
console.log(liczbySlownie(0.1));
console.log(liczbySlownie(1000000));
console.log(liczbySlownie(15));
