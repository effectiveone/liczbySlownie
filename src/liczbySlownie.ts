export interface Odmiana {
  jeden: string;
  dwa: string;
  piec: string;
}

export interface Konfiguracja {
  odmiana: Odmiana;
  separator: string;
  groszeSeparator: string;
}

const domyslnaKonfiguracja: Konfiguracja = {
  odmiana: {
    jeden: '',
    dwa: 'y',
    piec: '',
  },
  separator: ' ',
  groszeSeparator: ' i ',
};

const jednosci: string[] = [
  '',
  'jeden',
  'dwa',
  'trzy',
  'cztery',
  'pięć',
  'sześć',
  'siedem',
  'osiem',
  'dziewięć',
];
const nastki: string[] = [
  'dziesięć',
  'jedenaście',
  'dwanaście',
  'trzynaście',
  'czternaście',
  'piętnaście',
  'szesnaście',
  'siedemnaście',
  'osiemnaście',
  'dziewiętnaście',
];
const dziesiatki: string[] = [
  '',
  '',
  'dwadzieścia',
  'trzydzieści',
  'czterdzieści',
  'pięćdziesiąt',
  'sześćdziesiąt',
  'siedemdziesiąt',
  'osiemdziesiąt',
  'dziewięćdziesiąt',
];
const setki: string[] = [
  '',
  'sto',
  'dwieście',
  'trzysta',
  'czterysta',
  'pięćset',
  'sześćset',
  'siedemset',
  'osiemset',
  'dziewięćset',
];

export class LiczbySlownie {
  private konfiguracja: Konfiguracja;

  constructor(konfiguracja: Partial<Konfiguracja> = {}) {
    this.konfiguracja = { ...domyslnaKonfiguracja, ...konfiguracja };
  }

  public liczbySlownie(liczba: number): string {
    if (liczba === 0) return 'zero';

    const jestUjemna = liczba < 0;
    let reszta = Math.abs(liczba);
    const czesci: string[] = [];

    if (reszta >= 1000000000) {
      const miliardy = Math.floor(reszta / 1000000000);
      const miliardySlownie = this.przetworzGrupe(miliardy, 'miliard', 'miliardy', 'miliardów');
      if (miliardySlownie) czesci.push(miliardySlownie);
      reszta %= 1000000000;
    }

    if (reszta >= 1000000) {
      const miliony = Math.floor(reszta / 1000000);
      const milionySlownie = this.przetworzGrupe(miliony, 'milion', 'miliony', 'milionów');
      if (milionySlownie) czesci.push(milionySlownie);
      reszta %= 1000000;
    }

    if (reszta >= 1000) {
      const tysiace = Math.floor(reszta / 1000);
      const tysiaceSlownie = this.przetworzGrupe(tysiace, 'tysiąc', 'tysiące', 'tysięcy');
      if (tysiaceSlownie) czesci.push(tysiaceSlownie);
      reszta %= 1000;
    }

    if (reszta > 0) {
      const jednostkiSlownie = this.przetworzGrupe(reszta, '', '', '');
      if (jednostkiSlownie) czesci.push(jednostkiSlownie);
    }

    let wynik = czesci.filter(Boolean).join(this.konfiguracja.separator);
    if (jestUjemna) {
      wynik = 'minus' + (wynik ? this.konfiguracja.separator + wynik : '');
    }
    return wynik;
  }

  public kwotaSlownie(kwota: number): string {
    const jestUjemna = kwota < 0;
    const zlote = Math.floor(Math.abs(kwota));
    const grosze = Math.round((Math.abs(kwota) - zlote) * 100);

    let wynik = '';

    if (jestUjemna) {
      wynik += 'minus ';
    }

    wynik += this.liczbySlownie(zlote) + ' ' + this.odmianaZlotych(zlote);

    if (grosze > 0) {
      wynik +=
        this.konfiguracja.groszeSeparator +
        this.liczbySlownie(grosze) +
        ' ' +
        this.odmianaGroszy(grosze);
    }

    return wynik;
  }

  private przetworzGrupe(liczba: number, forma1: string, forma2: string, forma5: string): string {
    if (liczba === 0) return '';

    const setkiIndex = Math.floor(liczba / 100);
    const dziesiatkiIndex = Math.floor((liczba % 100) / 10);
    const jednosciIndex = liczba % 10;
    const liczbaJednosci = liczba % 100;

    const czesci: string[] = [];

    // Dodaj setki
    if (setkiIndex > 0) {
      czesci.push(setki[setkiIndex]);
    }

    // Dodaj dziesiątki i jedności
    if (liczbaJednosci >= 10 && liczbaJednosci <= 19) {
      czesci.push(nastki[liczbaJednosci - 10]);
    } else {
      if (dziesiatkiIndex > 0) {
        czesci.push(dziesiatki[dziesiatkiIndex]);
      }
      if (jednosciIndex > 0 || (liczba === 1 && forma1 === 'tysiąc')) {
        czesci.push(jednosci[jednosciIndex]);
      }
    }

    const liczebnik = czesci.filter(Boolean).join(this.konfiguracja.separator);
    const odmiana = this.odmianaLiczby(liczba, forma1, forma2, forma5);

    if (!odmiana) return liczebnik;
    if (!liczebnik) return odmiana;
    if (liczba === 1 && forma1 === 'tysiąc') return odmiana;

    return liczebnik + this.konfiguracja.separator + odmiana;
  }

  private odmianaLiczby(liczba: number, forma1: string, forma2: string, forma5: string): string {
    if (!forma1 && !forma2 && !forma5) return '';
    if (liczba === 1) return forma1;
    if (liczba % 100 >= 12 && liczba % 100 <= 14) return forma5;
    if (liczba % 10 >= 2 && liczba % 10 <= 4) return forma2;
    return forma5;
  }

  private odmianaZlotych(liczba: number): string {
    return this.odmianaLiczby(liczba, 'złoty', 'złote', 'złotych');
  }

  private odmianaGroszy(liczba: number): string {
    return this.odmianaLiczby(liczba, 'grosz', 'grosze', 'groszy');
  }
}
