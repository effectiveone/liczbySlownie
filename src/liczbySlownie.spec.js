import { liczbySlownie } from "./liczbySlownie";

describe("liczbySlownie", () => {
  test("konwertuje poprawnie liczby całkowite do 999", () => {
    expect(liczbySlownie(0)).toMatchSnapshot();
    expect(liczbySlownie(5)).toMatchSnapshot();
    expect(liczbySlownie(10)).toMatchSnapshot();
    expect(liczbySlownie(11)).toMatchSnapshot();
    expect(liczbySlownie(15)).toMatchSnapshot();
    expect(liczbySlownie(20)).toMatchSnapshot();
    expect(liczbySlownie(25)).toMatchSnapshot();
    expect(liczbySlownie(99)).toMatchSnapshot();
    expect(liczbySlownie(100)).toMatchSnapshot();
    expect(liczbySlownie(123)).toMatchSnapshot();
    expect(liczbySlownie(999)).toMatchSnapshot();
  });

  test("konwertuje poprawnie liczby całkowite do 999", () => {
    expect(liczbySlownie(0)).toBe("zero złotych");
    expect(liczbySlownie(5)).toBe("pięć złotych");
    expect(liczbySlownie(10)).toBe("dziesięć złotych");
    expect(liczbySlownie(11)).toBe("jedenaście złotych");
    expect(liczbySlownie(15)).toBe("piętnaście złotych");
    expect(liczbySlownie(20)).toBe("dwadzieścia złotych");
    expect(liczbySlownie(25)).toBe("dwadzieścia pięć złotych");
    expect(liczbySlownie(99)).toBe("dziewięćdziesiąt dziewięć złotych");
    expect(liczbySlownie(100)).toBe("sto złotych");
    expect(liczbySlownie(123)).toBe("sto dwadzieścia trzy złote");
    expect(liczbySlownie(999)).toBe(
      "dziewięćset dziewięćdziesiąt dziewięć złotych"
    );
  });

  test("konwertuje poprawnie liczby powyżej 999", () => {
    expect(liczbySlownie(1000)).toBe("jeden tysiąc złotych");
    expect(liczbySlownie(1500)).toBe("jeden tysiąc pięćset złotych");
    expect(liczbySlownie(2000)).toBe("dwa tysiące złotych");
    expect(liczbySlownie(2500)).toBe("dwa tysiące pięćset złotych");
    expect(liczbySlownie(9999)).toBe(
      "dziewięć tysięcy dziewięćset dziewięćdziesiąt dziewięć złotych"
    );
    expect(liczbySlownie(10000)).toBe("dziesięć tysięcy złotych");
    expect(liczbySlownie(10001)).toBe("dziesięć tysięcy jeden złoty");
    expect(liczbySlownie(10100)).toBe("dziesięć tysięcy sto złotych");
    expect(liczbySlownie(99999)).toBe(
      "dziewięćdziesiąt dziewięć tysięcy dziewięćset dziewięćdziesiąt dziewięć złotych"
    );
    expect(liczbySlownie(100000)).toBe("sto tysięcy złotych");
    expect(liczbySlownie(100001)).toBe("sto tysięcy jeden złoty");
    expect(liczbySlownie(101000)).toBe("sto jeden tysięcy złotych");
    expect(liczbySlownie(110000)).toBe("sto dziesięć tysięcy złotych");
    expect(liczbySlownie(200000)).toBe("dwieście tysięcy złotych");
    expect(liczbySlownie(300000)).toBe("trzysta tysięcy złotych");
    expect(liczbySlownie(400000)).toBe("czterysta tysięcy złotych");
    expect(liczbySlownie(500000)).toBe("pięćset tysięcy złotych");
    expect(liczbySlownie(600000)).toBe("sześćset tysięcy złotych");
    expect(liczbySlownie(700000)).toBe("siedemset tysięcy złotych");
    expect(liczbySlownie(800000)).toBe("osiemset tysięcy złotych");
    expect(liczbySlownie(900000)).toBe("dziewięćset tysięcy złotych");
    expect(liczbySlownie(1000000)).toBe("jeden milion złotych");
    expect(liczbySlownie(1000000000)).toBe("jeden miliard złotych");
    expect(liczbySlownie(2000000000)).toBe("dwa miliardy złotych");
  });
});

// testy dla groszy
describe("liczbySlownie dla groszy", () => {
  test("konwertuje poprawnie liczby z groszami", () => {
    expect(liczbySlownie(0.01)).toBe("jeden grosz");
    expect(liczbySlownie(0.05)).toBe("pięć groszy");
    expect(liczbySlownie(0.1)).toBe("dziesięć groszy");
    expect(liczbySlownie(0.11)).toBe("jedenaście groszy");
    expect(liczbySlownie(0.15)).toBe("piętnaście groszy");
    expect(liczbySlownie(0.2)).toBe("dwadzieścia groszy");
    expect(liczbySlownie(0.25)).toBe("dwadzieścia pięć groszy");
    expect(liczbySlownie(0.34)).toBe("trzydzieści cztery grosze");
    expect(liczbySlownie(0.99)).toBe("dziewięćdziesiąt dziewięć groszy");
  });
});
