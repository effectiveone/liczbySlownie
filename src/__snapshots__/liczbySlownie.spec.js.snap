// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`liczbySlownie konwertuje poprawnie liczby całkowite do 999 1`] = `"zero złotych"`;

exports[`liczbySlownie konwertuje poprawnie liczby całkowite do 999 2`] = `"pięć złotych"`;

exports[`liczbySlownie konwertuje poprawnie liczby całkowite do 999 3`] = `"dziesięć złotych"`;

exports[`liczbySlownie konwertuje poprawnie liczby całkowite do 999 4`] = `"jedenaście złotych"`;

exports[`liczbySlownie konwertuje poprawnie liczby całkowite do 999 5`] = `"piętnaście złotych"`;

exports[`liczbySlownie konwertuje poprawnie liczby całkowite do 999 6`] = `"dwadzieścia złotych"`;

exports[`liczbySlownie konwertuje poprawnie liczby całkowite do 999 7`] = `"dwadzieścia pięć złotych"`;

exports[`liczbySlownie konwertuje poprawnie liczby całkowite do 999 8`] = `"dziewięćdziesiąt dziewięć złotych"`;

exports[`liczbySlownie konwertuje poprawnie liczby całkowite do 999 9`] = `"sto złotych"`;

exports[`liczbySlownie konwertuje poprawnie liczby całkowite do 999 10`] = `"sto dwadzieścia trzy złote"`;

exports[`liczbySlownie konwertuje poprawnie liczby całkowite do 999 11`] = `"dziewięćset dziewięćdziesiąt dziewięć złotych"`;

exports[`liczbySlownie konwertuje poprawnie liczby do 1000 1`] = `""`;

exports[`liczbySlownie konwertuje poprawnie liczby do 1000 2`] = `"sto undefined"`;

exports[`liczbySlownie konwertuje poprawnie liczby do 1000 3`] = `"dwieście undefined"`;

exports[`liczbySlownie konwertuje poprawnie liczby do 1000 4`] = `"sto"`;

exports[`liczbySlownie konwertuje poprawnie liczby do 1000 5`] = `"sto pięćdziesiąt"`;

exports[`liczbySlownie konwertuje poprawnie liczby do 1000 6`] = `"dwieście"`;

exports[`liczbySlownie konwertuje poprawnie liczby do 1000 7`] = `"pięćset"`;

exports[`liczbySlownie konwertuje poprawnie liczby do 1000 8`] = `"sto"`;

exports[`liczbySlownie konwertuje poprawnie liczby do 1000 9`] = `"dwieście"`;

exports[`liczbySlownie konwertuje poprawnie liczby do 1000 10`] = `"pięćset"`;

exports[`liczbySlownie konwertuje poprawnie liczby do 1000 11`] = `"sto undefined tysiąc"`;

exports[`liczbySlownie liczbySlownie konwertuje poprawnie liczby do 1000 1`] = `"zero"`;

exports[`liczbySlownie liczbySlownie konwertuje poprawnie liczby do 1000 2`] = `" jeden"`;

exports[`liczbySlownie liczbySlownie konwertuje poprawnie liczby do 1000 3`] = `" dwa"`;

exports[`liczbySlownie liczbySlownie konwertuje poprawnie liczby do 1000 4`] = `" dziesięć"`;

exports[`liczbySlownie liczbySlownie konwertuje poprawnie liczby do 1000 5`] = `" piętnaście"`;

exports[`liczbySlownie liczbySlownie konwertuje poprawnie liczby do 1000 6`] = `" dwadzieścia"`;

exports[`liczbySlownie liczbySlownie konwertuje poprawnie liczby do 1000 7`] = `" pięćdziesiąt"`;

exports[`liczbySlownie liczbySlownie konwertuje poprawnie liczby do 1000 8`] = `" sto"`;

exports[`liczbySlownie liczbySlownie konwertuje poprawnie liczby do 1000 9`] = `" dwieście"`;

exports[`liczbySlownie liczbySlownie konwertuje poprawnie liczby do 1000 10`] = `" pięćset"`;

exports[`liczbySlownie liczbySlownie konwertuje poprawnie liczby do 1000 11`] = `" jeden tysiąc"`;
